src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1186,132 @@ void Client::HandleTimer(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE
+void Client::HandleTcpTimer(void)
+{
+    // If message wasn't received in timeout time close connection.
+    mEndpoint.SendEndOfStream();
+    mTcpInitialized = false;
+}
+
+void Client::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mEndpoint.SendByReference(mSendLink, 0);
```

Comment:
@abtink What you observed is correct. You need to wait for the "Established" callback before sending data on the connection.

This isn't fundamental and it can be changed, though I'd like to test it first. The original FreeBSD code supported "implied connect" --- if you call `sendto` on a socket that isn't connected, it will establish the connection and then deliver the data. Adding data to the socket before the connection is established could be written to follow those same code paths.

With TCP Fast Open (currently in progress) this will no longer be an issue.

Reply:
Thanks @samkumar. For now here in this PR we can keep this code as is.

I agree I think would be good (make it easier for user of API) if we can allow "implied connect" model (we can enhance it later).